### PR TITLE
boards: arm64: Add support for Intel SoC FPGA Agilex5 platform.

### PR DESCRIPTION
**Pull request by gdengi**
The Intel Agilex® 5 FPGA product family extends the innovations of the Intel Agilex® FPGA portfolio to midrange FPGA applications.
Added support for Intel SoC FPGA Agilex5 platform in the Zephyr with the following sub-modules and updates.

- Add 'intel_socfpga_agilex5_socdk' board
- Add 'agilex5' soc to Intel Agilex soc series
- Reset manager for Intel SoC FPGA Agilex5 platform
- Clock manager for Intel SoC FPGA Agilex5 platform
- UART, add an optional module reset line functionality
- PSCI, add support for multiple PSCI versions 0.2v and 1.1v
- Custom shell module configurations added as required by the Intel Agilex platforms.

### CODE REVIEW

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (povergoing) [inline boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:43]: Can you elaborate on why you set 25000000 in soc as a default value, but here you force it as 400000000

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (povergoing) [inline boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:5]: What is the `Intel Arm Trusted Firmware`, it's an Intel fork of ATF or ?

File: boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst
Comment (povergoing) [inline boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst:6]: It would be better to put more details like other boards did?

File: drivers/clock_control/clock_control_agilex5.c
Comment (povergoing) [inline drivers/clock_control/clock_control_agilex5.c:73]: Would you like to fix the alignments here?

**Review by carlocaione — COMMENTED** [review]
In general looks good. Just a couple of comments.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (fabiobaltieri) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:40]: Ideally... I'll hack it in my test PR and see if anything else complains, hang in there.

Comment (fabiobaltieri) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:40]: Run result: https://github.com/zephyrproject-rtos/zephyr/pull/58388/checks?check_run_id=14941982543, these were all failing already so yes - I think it would be ok to anticipate `RESET_INIT_PRIORITY` to 35 system wide. Maybe do a separate PR for it?

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.c
Comment (ickochar) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.c:25]: Any alternative suggested for such macros ?

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:5]: Yes, it is - where we have support for Intel SoC FPGA product lines.
And we are using the ATF as a boot loader to Zephyr.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:40]: So, shall I keep it here as of now as part this commit?

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.c
Comment (carlocaione) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.c:25]: just move `const struct config = dev->config` directly in place.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (carlocaione) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:40]: can you open a different PR to fix that?

Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:40]: Sure, will do. Thank you

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.h
Comment (gdengi) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.h]: Updated.

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.c
Comment (gdengi) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.c]: Updated to _const struct config = dev->config_ style and removed DEV_CFG macro.

File: drivers/clock_control/clock_control_agilex5.c
Comment (gdengi) [reply drivers/clock_control/clock_control_agilex5.c]: Updated.

File: boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst:6]: Will add shortly.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig]: Its a mistake. Removed from here, and added 400000000 in soc as a default value.

**Review by carlocaione — DISMISSED** [review]

File: drivers/serial/uart_ns16550.c
Comment (dcpleung) [inline drivers/serial/uart_ns16550.c:237]: Please use `DT_ANY_INST_HAS_PROP_STATUS_OKAY` to enable reset related stuff. If there is no reset controller attached (which is the majority of cases), there is no need to include all the added code and data.

**Review by dcpleung — CHANGES_REQUESTED** [review]
Also, could you use `samples/subsys/shell/shell_module/` instead of creating a new sample? The `shell_module` should also provide a shell which you can use to query devices.

**Comment by gdengi** [comment]
> Also, could you use `samples/subsys/shell/shell_module/` instead of creating a new sample? The `shell_module` should also provide a shell which you can use to query devices.

We wanted to keep simple and light CLI application for Intel SoC FPGA products line and that's where we introduce this application.
Sure, will try out the suggested sample and test. Will keep you updated. Thank you.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig]: @carlocaione  @fabiobaltieri  FYI, created a separate PR for this change. Thank you.
https://github.com/zephyrproject-rtos/zephyr/pull/60412

File: boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst:6]: Updated with all the details. Thank you.

File: boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig]: Corrected.

**Review by carlocaione — DISMISSED** [review]

File: drivers/serial/uart_ns16550.c
Comment (gdengi) [reply drivers/serial/uart_ns16550.c:262]: Updated the driver with DT_ANY_INST_HAS_PROP_STATUS_OKAY. Thank you.

**Comment by gdengi** [comment]
Hello @carlocaione  

All the review comments are addressed and all the required CI checks are passed.
Kindly let me know your inputs on the further steps.

Thank you.

Regards,
Girisha Dengi

**Review by carlocaione — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]

File: drivers/serial/uart_ns16550.c
Comment (fabiobaltieri) [inline drivers/serial/uart_ns16550.c:605]: Drop the initializer.

File: drivers/pm_cpu_ops/psci_shell.c
Comment (fabiobaltieri) [inline drivers/pm_cpu_ops/psci_shell.c:18]: drop the initializer

File: drivers/pm_cpu_ops/psci_shell.c
Comment (fabiobaltieri) [inline drivers/pm_cpu_ops/psci_shell.c:33]: same

File: drivers/pm_cpu_ops/psci_shell.c
Comment (fabiobaltieri) [inline drivers/pm_cpu_ops/psci_shell.c:48]: same

File: CODEOWNERS
Comment (fabiobaltieri) [inline CODEOWNERS:1]: Does `MAINTAINERS.yml` needs to be updated as well?

File: soc/arm64/intel_socfpga/agilex5/Kconfig.defconfig.agilex5
Comment (ickochar) [inline soc/arm64/intel_socfpga/agilex5/Kconfig.defconfig.agilex5:13]: update to value above 242

File: CODEOWNERS
Comment (gdengi) [reply CODEOWNERS:1]: Updated

File: drivers/pm_cpu_ops/psci_shell.c
Comment (gdengi) [reply drivers/pm_cpu_ops/psci_shell.c]: Updated

File: drivers/pm_cpu_ops/psci_shell.c
Comment (gdengi) [reply drivers/pm_cpu_ops/psci_shell.c]: Updated

File: drivers/pm_cpu_ops/psci_shell.c
Comment (gdengi) [reply drivers/pm_cpu_ops/psci_shell.c]: Updated

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.c
Comment (gdengi) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.c]: Updated as suggested. Thank you

File: drivers/serial/uart_ns16550.c
Comment (gdengi) [reply drivers/serial/uart_ns16550.c]: Updated

File: drivers/serial/uart_ns16550.c
Comment (gdengi) [reply drivers/serial/uart_ns16550.c]: Updated

File: drivers/serial/uart_ns16550.c
Comment (gdengi) [reply drivers/serial/uart_ns16550.c:613]: Updated

File: drivers/pm_cpu_ops/pm_cpu_ops_psci.c
Comment (gdengi) [reply drivers/pm_cpu_ops/pm_cpu_ops_psci.c]: Updated

File: drivers/serial/uart_ns16550.c
Comment (gdengi) [reply drivers/serial/uart_ns16550.c]: Updated

File: dts/arm64/intel/intel_socfpga_agilex5.dtsi
Comment (gdengi) [reply dts/arm64/intel/intel_socfpga_agilex5.dtsi]: Updated

File: drivers/clock_control/clock_control_agilex5_ll.c
Comment (gdengi) [reply drivers/clock_control/clock_control_agilex5_ll.c]: Updated

File: include/zephyr/dt-bindings/reset/intel_socfpga_reset.h
Comment (gdengi) [reply include/zephyr/dt-bindings/reset/intel_socfpga_reset.h]: Updated

File: drivers/reset/reset_intel_socfpga.c
Comment (gdengi) [reply drivers/reset/reset_intel_socfpga.c]: Updated

File: drivers/reset/reset_intel_socfpga.c
Comment (gdengi) [reply drivers/reset/reset_intel_socfpga.c]: Updated

**Comment by gdengi** [comment]
@fabiobaltieri Addressed all the review comments. Kindly take a look. Thank you.

File: boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst
Comment (gdengi) [reply boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst]: Updated

File: soc/arm64/intel_socfpga/agilex5/Kconfig.defconfig.agilex5
Comment (gdengi) [reply soc/arm64/intel_socfpga/agilex5/Kconfig.defconfig.agilex5:13]: Updated now as minimal required for initial Agilex5 BSP. 
Will update this as and when required drivers are implemented.

**Review by ickochar — APPROVED** [review]
Looks good to me!

**Review by carlocaione — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Comment by gdengi** [comment]
Hello @carlocaione, 

All the required approvals are in place - kindly take a look and help us to merge.
Thank you.

**Comment by carlocaione** [comment]
> All the required approvals are in place - kindly take a look and help us to merge. Thank you.

Unfortunately I don't have that power. But the mighty @fabiobaltieri has it.

**Comment by gdengi** [comment]
@dcpleung Kindly review and provide your feedback. Thank you.

**Comment by gdengi** [comment]
> > All the required approvals are in place - kindly take a look and help us to merge. Thank you.
> 
> Unfortunately I don't have that power. But the mighty @fabiobaltieri has it.
> 
> Also @dcpleung needs to review this again.

Thank you @carlocaione for the info.

**Review by dcpleung — APPROVED** [review]

**Comment by gdengi** [comment]
@fabiobaltieri  All the review comments are addressed and all the approvals are in place.
Kindly take a look and help us to merge.

Thank you.

File: MAINTAINERS.yml
Comment (stephanosio) [inline MAINTAINERS.yml:2300]: @gdengi Hi. The collaborator added here, `chongteikheng`, does not seem to be a valid GitHub account. Please submit a patch to either update or remove the username from the MAINTAINERS file.

File: CODEOWNERS
Comment (stephanosio) [inline CODEOWNERS:204]: ditto

Comment (gdengi) [reply CODEOWNERS:204]: Sure, will update here also.

File: MAINTAINERS.yml
Comment (gdengi) [reply MAINTAINERS.yml:2300]: Hello @stephanosio, sure will submit a patch to update the account name here. Thank you.

Comment (gdengi) [reply MAINTAINERS.yml:2300]: Hello @stephanosio, FYI, created a PR for the same. https://github.com/zephyrproject-rtos/zephyr/pull/65190

File: CODEOWNERS
Comment (gdengi) [reply CODEOWNERS:204]: Created a PR for the same. https://github.com/zephyrproject-rtos/zephyr/pull/65190

**Pull request merged**
